### DISCUSSION TITLE
chocolatey.lib - Questions about apikeys and packing

### DISCUSSION THREAD

**Discussion by TheCakeIsNaOH** [chocolatey/choco#2161] [General]
I have a couple of questions about using `chocolatey.lib`.

1. Is there a way to use an apikey added to the installed `choco.exe`? I.E. use an apikey added with `choco.exe apikey -s url -k key`? I would like to be able to push without having to specify the apikey each time. 
2. Is there a way to get what URLs there are ApiKeys for from the installed `choco.exe`? So I can check if there is an apikey available for the source I want to push to.

3. Is there something special in the config required to pack packages? I get an error (see below). Using `choco.exe pack` works fine.

```
Attempting to build package from 'nextcloud-client.nuspec'.
tools\Nextcloud-3.1.0-x64_x64.msi
Exception calling "Run" with "0" argument(s): "Can not access a closed Stream."
At line:1 char:1
+ $chocoLib.Run()
+ ~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : ObjectDisposedException
```

Source code snippet:

```
Add-Type -Path ".\lib\chocolatey.dll"
Add-Type -Path ".\lib\net40-full\log4net.dll"
$chocoLib = [chocolatey.Lets]::GetChocolatey()
$chocoLib.Set({
    param($config)
    $config.CommandName = 'pack'
    $config.Input = "F:\github\chocolatey-packages\nextcloud-client\nextcloud-client.nuspec"
    $config.OutputDirectory = "F:\temp\"
})
$chocoLib.Run()
```

**Comment by gep13** [comment]
@TheCakeIsNaOH quick question before digging too deep on this...

The `chocolatey.dll` that you have in your lib folder there, where are you getting that from?

**Reply by TheCakeIsNaOH** [reply]
I'm getting from the nupkg on nuget.org. 
https://www.nuget.org/api/v2/package/chocolatey.lib/0.10.15

**Reply by gep13** [reply]
OK, cool, just wanted to check.  I you were using a version of the dll that you had built yourself some of the answers would have been easier to address, but if you are using the official one, I will have to do some digging.  Leave this with me, and I will see what I can come up with.

**Comment by TheCakeIsNaOH** [comment]
I am playing with pack more, and pack does work for some packages, but not others.

It seems to be correlated to having binaries in the package. Any binary over 10-15mb seems to have an issue, but all of the smaller packages work fine.

**Reply by TheCakeIsNaOH** [reply]
Ok, yep, definitely individual binary size. 

A 400mb package with of lots of copies a 7mb binary works fine, but a package with a single 10mb binary fails.

Also, this is hilarious:
![image](https://user-images.githubusercontent.com/38865330/102507973-76cf8080-404a-11eb-8ac3-5a0f382129c6.png)

**Reply by gep13** [reply]
On this one, I am not sure.  I am not aware of any differences here.  @ferventcoder can you add anything here?

**Comment by gep13** [comment]
Regarding this question...

> Is there a way to use an apikey added to the installed choco.exe? I.E. use an apikey added with choco.exe apikey -s url -k key? I would like to be able to push without having to specify the apikey each time.

Can you show the code that you are using to call the push command?  Have you configured the Sources property on the Configuration object?  The reason that I ask is the following:

https://github.com/chocolatey/choco/blob/773b1852d9de9ab628598a9b29fe7d0f9d04f668/src/chocolatey/infrastructure.app/commands/ChocolateyPushCommand.cs#L101-L109

It should be looking up the ApiKey for the configured source.

**Reply by TheCakeIsNaOH** [reply]
I am using:
```
$chocoLib.Set({
    param($config)
    $config.CommandName = 'push'
    $config.Sources = "http://<internal-ip>:8081/repository/choco-hosted/"
    $config.Input = "F:\path\to\spotify\1.1.44.538\spotify.1.1.44.538.nupkg"
    $config.Force = $true
})
$chocoLib.Run()
```

And I get the error:
`"An ApiKey was not found for 'http://<internal-ip>:8081/repository/choco-hosted/'. You must
either set an api key in the configuration or specify one with --api-key."`

However, pushing using choco.exe works using this command:
`choco push -s  "http://<internal-ip>:8081/repository/choco-hosted/" -f F:\path\to\spotify\1.1.44.538\spotify.1.1.44.538.nupkg`

And using chocolatey.lib works if I manually specify the apikey:
```
$chocoLib.Set({
    param($config)
    $config.CommandName = 'push'
    $config.Sources = "http://<internal-ip>:8081/repository/choco-hosted/"
    $config.Input = "F:\path\to\spotify\1.1.44.538\spotify.1.1.44.538.nupkg"
    $config.PushCommand.Key = "<key>"
    $config.Force = $true
})
```

**Comment by gep13** [comment]
Regarding this:

> Is there a way to get what URLs there are ApiKeys for from the installed choco.exe? So I can check if there is an apikey available for the source I want to push to.

That would simply be the output from running `choco apikey` right?

**Reply by TheCakeIsNaOH** [reply]
Correct, that output would be fine.

I tried to use 
```
Add-Type -Path ".\lib\chocolatey.dll"
Add-Type -Path ".\lib\net40-full\log4net.dll"
$chocoLib = [chocolatey.Lets]::GetChocolatey()
$chocoLib.Set({
    param($config)
    $config.CommandName = [chocolatey.infrastructure.app.domain.CommandNameType]::apikey
})
$chocoLib.Run()
```

But I get no output back.

**Comment by TheCakeIsNaOH** [comment]
@gep13 Have you had a chance to dig into this yet?
